### PR TITLE
Validation matrix

### PR DESCRIPTION
**Pull request by alishaahuja1996**

### CODE REVIEW

**Review by ianaf100 — APPROVED** [review]

File: lib/src/input_validation/validate_function.dart [hunk 1]
```diff
@@ -40,23 +39,151 @@ class ValidateFunction {
...
-      if(currentState is ErrorState) {
+      if (currentState is ErrorState) {
         valid = false;
       }
     }
 
     return valid;
   }
 
+  bool testMatrixFunction(String expression){
```
Comment (gwarchol) [inline lib/src/input_validation/validate_function.dart]: lets move this to a separate class, its pretty large and does something pretty separate

File: lib/src/input_validation/validate_function.dart [hunk 2]
```diff
@@ -40,23 +39,151 @@ class ValidateFunction {
...
+    for (int i=0; i < rowsMatrix.length; i++) {
+      List<String> colMatrix = rowsMatrix[i].split(",");
+      if (colSize != colMatrix.length)
+        return false;
+    }
+
+    return true;
+  }
+
+  List<String> _sanitizeMatrixInput(String input){
```
Comment (gwarchol) [inline lib/src/input_validation/validate_function.dart]: i made a class that adds padding you could most likely use

File: lib/src/input_validation/validate_function.dart [hunk 3]
```diff
@@ -40,23 +39,151 @@ class ValidateFunction {
...
+
+  bool isRowColSameSize(String matrix1, String matrix2){
+    List<String> rowsMatrix2 = matrix2.split("!");
+    List<String> rowsMatrix1 = matrix1.split("!");
+
+    //need to check if correct number of columns exist in a row for matrix
+    if(isValidColumns(rowsMatrix1) && isValidColumns(rowsMatrix2)) {
+      List<String> colMatrix1 = rowsMatrix1[0].split(",");
+
+      if (rowsMatrix2.length == colMatrix1.length)
```
Comment (gwarchol) [inline lib/src/input_validation/validate_function.dart]: you can just do return rowsMatrix2.length == colMatrix1.length; or assign it to a bool and return it, using the if else is a bit of unnecessary complexity

File: lib/src/input_validation/validate_function.dart [hunk 4]
```diff
@@ -40,23 +39,151 @@ class ValidateFunction {
...
+      return false;
+  }
+
+  bool isValidColumns(List<String> rowsMatrix){
+    int colSize = rowsMatrix[0].split(",").length;
+
+    for (int i=0; i < rowsMatrix.length; i++) {
+      List<String> colMatrix = rowsMatrix[i].split(",");
+      if (colSize != colMatrix.length)
+        return false;
```
Comment (gwarchol) [inline lib/src/input_validation/validate_function.dart]: try to allow loops to run to completion, it makes code more readable

**Review by gwarchol — CHANGES_REQUESTED** [review]

File: lib/src/input_validation/validate_function.dart (hunk 2 above)
Comment (alishaahuja1996) [reply lib/src/input_validation/validate_function.dart]: I am aware. I can't use it. It also adds spacing between ',' which I don't want and messes up my list split

File: lib/src/input_validation/validate_function.dart (hunk 1 above)
Comment (alishaahuja1996) [reply lib/src/input_validation/validate_function.dart]: will do

Comment (alishaahuja1996) [reply lib/src/input_validation/validate_function.dart]: done

File: lib/src/input_validation/validate_function.dart (hunk 3 above)
Comment (alishaahuja1996) [reply lib/src/input_validation/validate_function.dart]: done

File: lib/src/input_validation/validate_function.dart (hunk 4 above)
Comment (alishaahuja1996) [reply lib/src/input_validation/validate_function.dart]: I don't think it makes sense to go through a complete for loop if one item fails so leaving is as is.

**Pull request merged**